File: src/lib/query.ts
/** TypeScript mirror of python-backend/query.py — keep the two in sync.
 * Used to render filter chips client-side without a round trip. */

export interface ParsedQuery {
  text: string[];
  filters: Record<string, string[]>;
}

const KNOWN_KEYS = new Set([
  'person', 'people', 'place', 'type', 'is', 'folder', 'in',
  'after', 'before', 'on', 'year', 'label',
]);

/** Minimal shlex-like tokenizer supporting double quotes. */
function tokenize(raw: string): string[] {
  const tokens: string[] = [];
  let current = '';
  let inQuotes = false;
  for (const ch of raw) {
    if (ch === '"') {
      inQuotes = !inQuotes;
    } else if (ch === ' ' && !inQuotes) {
      if (current) tokens.push(current);
      current = '';
    } else {
      current += ch;
    }
  }
  if (current) tokens.push(current);
  return tokens;
}

export function parse(raw: string): ParsedQuery {
  const text: string[] = [];
  const filters: Record<string, string[]> = {};
  for (const token of tokenize(raw || '')) {
    const idx = token.indexOf(':');
    if (idx > 0) {
      const key = token.slice(0, idx).toLowerCase();
      const value = token.slice(idx + 1);
      if (key === 'year' && /^\d{4}$/.test(value)) {
        (filters.after ??= []).push(`${value}-01-01`);
        (filters.before ??= []).push(`${Number(value) + 1}-01-01`);
        continue;
      }
      if (KNOWN_KEYS.has(key) && value) {
        const normalized = key === 'people' ? 'person' : key === 'in' ? 'place' : key;
        (filters[normalized] ??= []).push(value);
        continue;
      }
    }
    text.push(token);
  }
  return { text, filters };
}

File: python-backend/query.py
"""Search query parser.

Grammar: free-text words plus ``key:value`` tokens (value may be quoted).
``year:`` is sugar for an after/before pair. Keys repeat (list semantics).
This parser is mirrored in TypeScript for the frontend's filter chips — keep
the two in sync.
"""

import shlex
from dataclasses import dataclass, field

KNOWN_KEYS = {
    "person", "people", "place", "type", "is", "folder", "in",
    "after", "before", "on", "year", "label",
}


@dataclass
class Query:
    text: list = field(default_factory=list)
    filters: dict = field(default_factory=dict)

    def get(self, key: str) -> list:
        return self.filters.get(key, [])


def parse(raw: str) -> Query:
    query = Query()
    try:
        tokens = shlex.split(raw or "")
    except ValueError:
        tokens = (raw or "").split()
    for token in tokens:
        if ":" in token:
            key, _, value = token.partition(":")
            key = key.lower()
            value = value.strip().strip('"')
            if key == "year" and value.isdigit() and len(value) == 4:
                query.filters.setdefault("after", []).append(f"{value}-01-01")
                query.filters.setdefault("before", []).append(f"{int(value) + 1}-01-01")
                continue
            if key in KNOWN_KEYS and value:
                if key in ("person", "people"):
                    key = "person"
                elif key in ("place", "in"):
                    key = "place"
                query.filters.setdefault(key, []).append(value)
                continue
        query.text.append(token)
    return query
